net/captivedetection/captivedetection.go

```diff
@@ -0,0 +1,226 @@
+	defer httpClient.CloseIdleConnections()
+
+	endpoints := availableEndpoints(preferredDERPRegionID)
+	if len(endpoints) == 0 {
+		logf("no endpoints available for captive portal detection")
+		return false
+	}
+
+	// To avoid false positives, we issue requests to a maximum of 6 captive portal detection URLs
+	// and only report a captive portal is present when at least 2 responses look like a captive
+	// portal.
+	foundCount := 0
+	for _, e := range endpoints[:6] {
+		found, err := checkCaptivePortalEndpoint(ctx, e, ifIndex, logf)
+		if err != nil {
+			logf("checkCaptivePortalEndpoint failed with endpoint %s: %v", e.URL, err)
+			continue
+		}
+		if found {
+			foundCount++
+		}
+		if foundCount == 2 {
+			return true
+		}
+	}
+	return false
+}
+
+// checkCaptivePortalEndpoint checks if the given Endpoint is a captive portal by making an HTTP request to the
+// given Endpoint URL and checking if the response looks like a captive portal.
+func checkCaptivePortalEndpoint(ctx context.Context, e Endpoint, ifIndex int, logf func(format string, a ...any)) (found bool, err error) {
+	req, err := http.NewRequestWithContext(ctx, "GET", e.URL, nil)
+	if err != nil {
+		return false, err
+	}
+
+	// Attach the Tailscale challenge header if the endpoint supports it. Not all captive portal detection endpoints
+	// support this, so we only attach it if the endpoint does.
+	if e.SupportsTailscaleChallenge {
+		if u, err := url.Parse(e.URL); err == nil {
+			// Note: the set of valid characters in a challenge and the total
+			// length is limited; see isChallengeChar in cmd/derper for more
+			// details.
+			chal := "ts_" + u.Host
+			req.Header.Set("X-Tailscale-Challenge", chal)
+		}
+	}
+
+	// Force the HTTP connection to use the given interface.
+	// Why? On Darwin, until you've accepted or closed the macOS captive portal alert,
+	// no default route interface is available, so all requests will fail unless we do this.
+	// TODO: investigate if this is also necessary on other platforms (Windows?)
+	d := &net.Dialer{
+		Control: func(network, address string, c syscall.RawConn) error {
+			return prepareRawConn(c, ifIndex, logf)
+		},
+	}
+	httpClient.Transport = &http.Transport{
+		DialContext: func(ctx context.Context, network, addr string) (net.Conn, error) {
+			return d.Dial(network, addr)
+		},
+	}
+
+	// Make the actual request, and check if the response looks like a captive portal or not.
+	r, err := httpClient.Do(req)
+	if err != nil {
+		return false, err
+	}
+	return e.responseLooksLikeCaptive(r, logf), nil
+}
+
+// responseLooksLikeCaptive checks if the given HTTP response matches the expected response for the Endpoint.
+func (e Endpoint) responseLooksLikeCaptive(r *http.Response, logf func(format string, a ...any)) bool {
+	defer r.Body.Close()
+
+	if r.StatusCode == e.StatusCode {
+		logf("captive portal check request returned expected status code: %d", r.StatusCode)
+		if e.ExpectedContent != "" {
```

Comment:
You could invert and then de-indent this:

```go
if e.ExpectedContent == "" {
	logf("captive portal check response: no need to check response body for %q", e.URL)
	return false
}

// [... the rest ...]

Reply:
much better now